### PR TITLE
Apollo + Authentication + graph.cool example

### PR DESCRIPTION
**Pull request by jesstelford**
Who should be in charge of setting up the demo site? https://next-with-apollo-auth.now.sh

This is an extention of the _[with Apollo](https://github.com/zeit/next.js/tree/master/examples/with-apollo#the-idea-behind-the-example)_ example:

> [Apollo](http://dev.apollodata.com) is a GraphQL client that allows you to easily query the exact data you need from a GraphQL server. In addition to fetching and mutating data, Apollo analyzes your queries and their results to construct a client-side cache of your data, which is kept up to date as further queries and mutations are run, fetching more results from the server.
> 
> In this simple example, we integrate Apollo seamlessly with Next by wrapping our *pages* inside a [higher-order component (HOC)](https://facebook.github.io/react/docs/higher-order-components.html). Using the HOC pattern we're able to pass down a central store of query result data created by Apollo into our React component hierarchy defined inside each page of our Next application.
> 
> On initial page load, while on the server and inside `getInitialProps`, we invoke the Apollo method,  [`getDataFromTree`](http://dev.apollodata.com/react/server-side-rendering.html#getDataFromTree). This method returns a promise; at the point in which the promise resolves, our Apollo Client store is completely initialized.
> 
> This example relies on [graph.cool](https://www.graph.cool) for its GraphQL backend.
> 
> *Note: Apollo uses Redux internally; if you're interested in integrating the client with your existing Redux store check out the [`with-apollo-and-redux`](https://github.com/zeit/next.js/tree/master/examples/with-apollo-and-redux) example.*

[graph.cool](https://www.graph.cool) can be setup with many different
[authentication providers](https://www.graph.cool/docs/reference/integrations/overview-seimeish6e/#authentication-providers), the most basic of which is [email-password authentication](https://www.graph.cool/docs/reference/simple-api/user-authentication-eixu9osueb/#email-and-password). Once email-password authentication is enabled for your graph.cool project, you are provided with 2 useful mutations: `createUser` and `signinUser`.

On loading each route, we perform a `user` query to see if the current visitor is logged in (based on a cookie, more on that in a moment). Depending on the query result, and the route, the user may be [redirected](https://github.com/zeit/next.js/blob/master/examples/with-apollo-auth/lib/redirect.js) to a different page.

When creating an account, both the `createUser` and `signinUser` mutations are executed on graph.cool, which returns a token that can be used to [authenticate the user for future requests](https://www.graph.cool/docs/reference/auth/authentication-tokens-eip7ahqu5o/). The token is stored in a cookie for easy access (_note: This may have security implications. Please understand XSS and JWT before deploying this to production_).

A similar process is followed when signing in, except `signinUser` is the only mutation executed.

It is important to note the use of Apollo's `resetStore()` method after signing in and signing out to ensure that no user data is kept in the browser's memory.

To get this example running locally, you will need to create a graph.cool
account, and provide [the `project.graphcool` schema](https://github.com/zeit/next.js/blob/master/examples/with-apollo-auth/project.graphcool).

### CODE REVIEW

**Comment by timneutkens** [comment]
@ads1018 could you review?

**Comment by ads1018** [comment]
brilliant <3. Was hoping someone would take the reins on a good Apollo auth example :)

@jesstelford could you lint your committed files with [standard](https://standardjs.com/)? If you use atom you can install the `atom-standard-formatter` package and use `ctrl-alt-f` to automatically format the current Javascript file your on.

**Comment by timneutkens** [comment]
Thanks for reviewing @ads1018. @jesstelford when travis is green (standard changes) I'll merge.

File: examples/with-apollo-auth/lib/init-apollo.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+// Polyfill fetch() on the server (used by apollo-client)
+if (!process.browser) {
+  global.fetch = fetch
+}
+
+function create (initialState, { getToken }) {
+
+  const networkInterface = createNetworkInterface({
+    uri: 'https://api.graph.cool/simple/v1/cj3h80ffbllm20162alevpcby',
```
Comment (notrab) [inline examples/with-apollo-auth/lib/init-apollo.js]: I'd recommend here that this doesn't use a specific project ID and uses an ENV variable or some instructions inside the README pointing people to use your schema with their CLI.

```
graphcool init --schema schema.graphql
```

File: examples/with-apollo-auth/lib/redirect.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import Router from 'next/router';
+
+export default (context, target) => {
+  if (context.res) {
```
Comment (juhaelee) [inline examples/with-apollo-auth/lib/redirect.js]: couldn't we do a `if (!process.browser)` here?

**Comment by timneutkens** [comment]
@jesstelford could you fix the linting errors?

**Review by timneutkens — CHANGES_REQUESTED** [review]
Linting errors

**Review by timneutkens — APPROVED** [review]
I fixed the linting errors :)

**Comment by timneutkens** [comment]
Thanks!

**Comment by mnsrv** [comment]
Thanks @jesstelford and @timneutkens for this great example!

Can we change using FormData (because it's [not supported](https://developer.mozilla.org/en-US/docs/Web/API/FormData/get) in Safari and IE) for something else? For my project I use values from input directly like `event.target.elements['email'].value` but this is dirty. May be we can use input values from state? or something else.

Thank you again!

**Comment by marktani** [comment]
Just saw this, this is amazing! :tada:

**Comment by jesstelford** [comment]
Sorry for PRing and Running! (life happened). Thank you so much for everyone's support and for getting this merged - hopefully it inspires more folks to get started with GraphQL :tada:

**Comment by niwaa** [comment]
Some issue with this with-apollo-auth example.  

When I try to access a page when NOT authenticated, I get errors "Network error: document is not defined", which is coming from with-data.js, "document.cookie" in the parseCookies function. 

Not sure to understand from this repo example how to handle a page that doesn't need the user to be authenticated. 

I have a bunch of pages on the app that should be public, with authenticated users having edit privileges.

**Comment by DevSpeak** [comment]
Any chance we could get this as a HOC?

**Pull request merged**